cilium/CFP-27752-operator-manages-cilium-identities.md

```diff
@@ -0,0 +1,214 @@
+
+Pod startup latency will be reduced when a CID for pod’s labels doesn’t exist yet. CID creation is time consuming mainly because of waiting for kube-apiserver to create the CID resource in Etcd. With operator managing CIDs, CID creation will happen as soon as Pod object is created, before it's even scheduled. Therefore, Agents will always have CID in the watcher store at the time of pod startup.
+
+## Pod creation process
+
+CID assignment paths on pod creation:
+
+1. Regular path: Pod created -> CID created (Operator) -> Pod scheduled -> Cilium CNI -> CID assigned
+2. Missing global ID path: Pod created -> Pod scheduled -> Cilium CNI -> Fail to find CID in watcher Store -> Retry/Fail CNI -> Pod creation is retried by kubelet.
+
+The diagram below covers high level actions that happen when a pod is created. From pod scheduled to pod ready, and to CEP and CID data being received by all nodes in the cluster. CEP batching into CES is included, but that’s an optional step based on configuration. Comparison before (red arrows) and after (blue arrows) changes.
+
+![](../cilium/images/CFP-27752-pod-creation-process.png)
+
+
+CEP creation in parallel triggers its endpoint regeneration on the host node and CEP propagation to all other nodes.
+
+Endpoint regeneration happens when:
+
+* An endpoint is created and updated, including pod and namespace label changes (security identity change).
+* CID is created or deleted - all Agents regenerate all of their endpoints.
+* Network policies are changed.
+
+IPCache is updated when remote CEPs/CESs are received.
+
+## Label changes
+
+
+### Pod
+
+Pod/Deployment label changes result in pod recreation, which follows the same pod creation path previously described.
+
+
+### Namespace
+
+When namespace labels change, Operator and Agents get the update at the same time. Agents wait until Operator processes the namespace label change update and creates all the new CIDs. Agents keep retrying to match CIDs for all their local endpoints until they are received through the CID watcher.
+
+
+## Simultaneous CID management compatibility
+
+The only problem with simultaneous CID management (both Operator and Agents are managing CIDs) is when namespace labels are changed. The only bad thing that happens is that there is CID duplication. This is an existing problem already, that doesn't cause disruption.
+
+Simultaneous CID management is only expected to happen during upgrade from Agents managing CIDs to Operator managing CIDs.
+
+
+## CID deduplication
+
+Operator will not create duplicate CIDs, but it will allow for their existence. In case other sources, like Agents, are making duplicate CIDs, Operator will be compatible with tracking multiple CIDs. This is needed to make Operator compatible with Agents managing CIDs at the same time.
+
```

Comment:
Is there a plan for the operator to eventually de-duplicate the CIDs? (doesn't have to be part of this CFP, but it could be mentioned as a future milestone)

Reply:
Yes, but it's not so simple, unless we want to add this logic to Agents.

Duplicated CIDs are CIDs that have the same label set.

There is logic in the CID controller to pick the first CID that it finds in the CID watcher store that matches a new pod, and always associate all pods with that label set to that single CID. It's already implemented in the method `handleStoreCIDMatch()` - https://github.com/cilium/cilium/blob/main/operator/pkg/ciliumidentity/reconciler.go#L397

However, this doesn’t entirely solve the problem, because as long as multiple CIDs with the same label set exist, Agents will be able to assign any of them to new CEPs. Therefore, it can take a significant amount of time and pod churn to completely de-duplicate CIDs, because it will be random. This is especially true with the legacy CID GC, which is taking 30 minutes to clean up CIDs.

With CID GC enhancements, when it’s moved to CID controller, the delete delay will be much smaller (e.g. 30 seconds), which will make de-duplication more rapid.

Reply:
I see yeah. I was thinking of somehow having some ordering metric so that agents could listen for CIDs and detect duplicates, then all independently pick one "original" instance of the duplicate CID to switch endpoints over to. Once the agent picks this "original" CID, it would actively change local endpoints over to use that one instead of the duplicate. If all agents independently do this, they could all align on using the same CID in a distributed way, then they would each remove references to the other duplicate CIDs. This would free the duplicates up for release to the pool (without even requiring Pod churn).

Still though this problem is independent and has a bunch of nuance to it so maybe we can tackle that separately from this CFP if you think it's worth pursuing.

Reply:
I think better tackle it separately from this CFP.

I added to the paragraph the following:

Completely deduplicating CIDs could be done with additional changes in Agent. One option is to have Agents always pick CIDs with the lowest integer value in the name, that matches the label set. This way all Agents would always pick the same CID for the same label set. However, this is only a suggestion, and a proper solution will be covered separately, outside this CFP.